File: src/CSharp.Repository/Services/ServiceConfiguration.cs
using CSharp.Forms.Data;
using CSharp.Repository.Interfaces;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using System;
using System.Collections.Generic;
using System.Configuration;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace CSharp.Repository.Services
{
    public static class ServiceConfiguration
    {
        private static IServiceProvider serviceProvider { get; set; }

        /// <summary>
        /// Adds the scoped services into the <see cref="ServiceCollection"/>
        /// </summary>
        /// <param name="forms"></param>
        /// <param name="Configuration"></param>
        public static void ConfigureServices(List<Type> forms, IConfiguration configuration)
        {
            var services = new ServiceCollection();
            forms?.ForEach(x => services.AddTransient(x));
            services.AddSingleton(configuration);
            services.AddDbContext<DataContext>(options => options.UseInMemoryDatabase(configuration.GetConnectionString("WinformsDatabase")));
            services.AddScoped<IHelloWorkRepository, HelloWorkRepository>();
            services.AddScoped<IUsersRepository, UsersRepository>();
            serviceProvider = services.BuildServiceProvider();
        }

        /// <summary>
        /// Returns a new instance of the scoped service associated with type <typeparamref name="T"/>
        /// </summary>
        public static T GetProviderService<T>() where T : class
        {
            if (serviceProvider == null) return null;
            return (T)serviceProvider.GetService(typeof(T));
        }

        /// <summary>
        /// Creates an instance of the form
        /// </summary>
        public static T CreateForm<T>() where T : Form
        {
            if (serviceProvider == null) return null;
            return (T)serviceProvider.GetService(typeof(T));
        }
    }
}

File: src/VB.Repository/Services/ServiceConfiguration.vb

Imports System.Configuration
Imports System.Windows.Forms
Imports Microsoft.EntityFrameworkCore
Imports Microsoft.Extensions.Configuration
Imports Microsoft.Extensions.DependencyInjection
Imports Microsoft.Extensions.Logging
Imports VB.Forms.Data.VB.Forms.Data
Imports VB.Repository.VB.Repository.Interfaces

Namespace VB.Repository.Services

    ''' <summary>
    ''' A module containing the <see cref="ServiceCollection"/> used for dependency injection, throughout the application
    ''' </summary>
    Public Module ServiceConfiguration

        Private Property ServiceProvider As IServiceProvider

        ''' <summary>
        ''' Adds the scoped services into the <see cref="ServiceCollection"/>
        ''' </summary>
        ''' <param name="forms"></param>
        Public Sub ConfigureServices(forms As List(Of Type))

            Dim builder = New ConfigurationBuilder().
            AddJsonFile("appsettings.json", [optional]:=True, reloadOnChange:=True)

            Dim configuration = builder.Build()
            Dim services As New ServiceCollection()

            'find all forms in the solution and add them to the ServiceCollection

            forms?.ForEach(Sub(form)
                               services.AddTransient(form)
                           End Sub)
            services.
            AddSingleton(Of IConfiguration)(configuration).
            AddDbContext(Of DataContext)(Function(options As DbContextOptionsBuilder) options.UseInMemoryDatabase(configuration.GetConnectionString("WinformsDatabase"))).
            AddSingleton(Of IFormFactory, FormFactory)().
            AddScoped(Of IHelloWorkRepository, HelloWorkRepository)().
            AddScoped(Of IUsersRepository, UsersRepository)().
            AddLogging(Function(configure) configure.AddDebug())
            Dim serviceProvider = services.BuildServiceProvider()
            SetServiceProvider(serviceProvider)
        End Sub

    End Module

End Namespace


